File: apps/driver-android/src/main/kotlin/lk/mageride/driver/support/SupportSheets.kt
package lk.mageride.driver.support

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AttachFile
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuAnchorType
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import kotlinx.coroutines.launch
import lk.mageride.driver.R
import lk.mageride.driver.capture.readImage
import lk.mageride.driver.jobs.ScheduleLabels
import lk.mageride.driver.ui.Symbols
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.SectionLabel
import lk.mageride.driver.ui.component.StatusPill
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.query.TripSummary
import lk.mageride.shared.data.models.support.TicketDetail
import lk.mageride.shared.data.models.support.TicketEvent

/**
 * SCR-DA-033's three overlays, hosted together because at most one is ever up.
 *
 * The raise-ticket sheet is the wireframe's own SCR-DA-033a; the article and thread sheets are
 * US-16.1's *"+ detail"* and US-16.2's *"track ticket"*, which D2' names and the wireframe does not
 * draw a screen for. Both are `ModalBottomSheet` rather than routes for that reason — adding two
 * destinations the team-approved baseline has no frame for would be a deviation, and a sheet over
 * the list the driver tapped is the least of it.
 */
@Composable
internal fun SupportSheets(state: SupportState, viewModel: SupportViewModel) {
    when (state.sheet) {
        SupportSheet.RAISE_TICKET -> RaiseTicketSheet(state = state, viewModel = viewModel)

        SupportSheet.TICKET_THREAD -> TicketThreadSheet(
            ticket = state.ticket,
            onDismiss = viewModel::closeSheet,
        )

        SupportSheet.FAQ_ARTICLE -> ArticleSheet(
            title = state.article?.title,
            body = state.article?.body,
            onDismiss = viewModel::closeSheet,
        )

        null -> Unit
    }
}

/**
 * **SCR-DA-033a · raise a ticket** — an M3 modal bottom sheet (US-16.2).
 *
 * The wireframe's four controls, in its order: **Issue description**, the **Related trip**
 * dropdown, *"📎 Attach a screenshot"* and **Submit ticket**. The title changes with the category —
 * the same sheet is the daily-fee refund request (US-9.23) — because both post the same operation
 * and only `category` differs.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun RaiseTicketSheet(state: SupportState, viewModel: SupportViewModel) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    // The system photo picker: no READ_MEDIA_IMAGES, no storage permission on any API level this
    // app supports, and the grant dies with the pick — the same contract SCR-DA-005's gallery
    // fallback uses. `readImage` takes the bytes now, because the Uri does not outlive the process.
    val pickScreenshot = rememberLauncherForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
        if (uri != null) {
            scope.launch { viewModel.onScreenshotPicked(readImage(context, uri, SCREENSHOT_FILE_NAME)) }
        }
    }

    ModalBottomSheet(onDismissRequest = viewModel::closeSheet, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            Text(
                text = stringResource(
                    if (state.isRefundRequest) R.string.support_refund_title else R.string.support_raise_ticket,
                ),
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )

            OutlinedTextField(
                value = state.description,
                onValueChange = viewModel::onDescriptionChange,
                modifier = Modifier.fillMaxWidth(),
                label = { Text(text = stringResource(R.string.support_issue_label)) },
                placeholder = { Text(text = stringResource(R.string.support_issue_hint)) },
                minLines = MIN_DESCRIPTION_LINES,
                singleLine = false,
            )

            SectionLabel(text = stringResource(R.string.support_related_trip))
            TripDropdown(state = state, onSelect = viewModel::onTripSelected)

            OutlinedButton(onClick = { pickScreenshot.launch(imageOnly()) }, modifier = Modifier.fillMaxWidth()) {
                Icon(imageVector = Icons.Outlined.AttachFile, contentDescription = null)
                Text(
                    text = stringResource(
                        if (state.screenshot == null) {
                            R.string.support_attach
                        } else {
                            R.string.support_attached
                        },
                    ),
                    modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                )
            }

            MageRideCta(
                label = stringResource(R.string.support_submit_ticket),
                onClick = viewModel::submit,
                enabled = state.canSubmit,
                loading = state.submitting,
            )
        }
    }
}

/**
 * The wireframe's *"DRV-22011-0617 · Galle Face → Nugegoda ▾"*.
 *
 * The id in the wireframe is a driver-and-date composite this platform does not mint (the same
 * finding C074 recorded about `DRV-22011`), so the row is the **route and the day** — which is what
 * a driver recognises a trip by, and what the support agent will search on. Optional: a fee charged
 * when the app crashed on Go Online is not about a trip at all.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TripDropdown(state: SupportState, onSelect: (String?) -> Unit, modifier: Modifier = Modifier) {
    var expanded by remember { mutableStateOf(false) }
    val selected = state.trips.firstOrNull { it.tripId == state.tripId }

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = !expanded },
        modifier = modifier.fillMaxWidth(),
    ) {
        OutlinedTextField(
            value = selected?.let { tripLabel(it) } ?: stringResource(R.string.support_trip_none),
            onValueChange = {},
            modifier = Modifier
                .fillMaxWidth()
                // `ExposedDropdownMenuAnchorType` is the Material3 1.4 spelling; the argument-less
                // `menuAnchor()` is deprecated there. See apps/driver-android/CLAUDE.md.
                .menuAnchor(ExposedDropdownMenuAnchorType.PrimaryNotEditable),
            readOnly = true,
            singleLine = true,
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            DropdownMenuItem(
                text = { Text(text = stringResource(R.string.support_trip_none)) },
                onClick = {
                    onSelect(null)
                    expanded = false
                },
            )
            state.trips.forEach { trip ->
                DropdownMenuItem(
                    text = { Text(text = tripLabel(trip)) },
                    onClick = {
                        onSelect(trip.tripId)
                        expanded = false
                    },
                )
            }
        }
    }
}

/** One ticket and its whole conversation (US-16.2). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TicketThreadSheet(ticket: TicketDetail?, onDismiss: () -> Unit) {
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            if (ticket == null) {
                Text(
                    text = stringResource(R.string.support_loading),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                return@Column
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                Text(
                    text = categoryLabel(ticket.category),
                    modifier = Modifier.weight(1f),
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                StatusPill(
                    label = stringResource(SupportLabels.status(ticket.status)),
                    tone = SupportLabels.tone(ticket.status),
                )
            }

            Text(
                text = ticket.description,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )

            // Oldest first, as the contract sends it: a thread read bottom-up is a thread nobody
            // reads. `assigned` entries are skipped — who is handling a complaint is not the
            // driver's to see, which is the contract's own rule rather than this screen's.
            ticket.thread.forEach { event -> ThreadEntry(event = event) }
        }
    }
}

/** One `TicketEvent` — what happened, and the agent's words when there are any. */
@Composable
private fun ThreadEntry(event: TicketEvent, modifier: Modifier = Modifier) {
    val label = SupportLabels.event(event.kind) ?: return

    Column(modifier = modifier.fillMaxWidth()) {
        Text(
            text = stringResource(label),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        event.body?.takeIf(String::isNotBlank)?.let { body ->
            Text(
                text = body,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
    }
}

/** One FAQ article's body (US-16.1). Server-rendered in the driver's own language (D-26). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ArticleSheet(title: String?, body: String?, onDismiss: () -> Unit) {
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Text(
                text = title ?: stringResource(R.string.support_loading),
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
                textAlign = TextAlign.Start,
            )
            body?.let {
                // Markdown as written. The app ships no renderer and support-svc's articles are
                // short prose; a half-implemented one that swallowed `#` would be worse than none.
                Text(
                    text = it,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

/**
 * *"12 Jun · Galle Face → Nugegoda"*.
 *
 * The date is read in **Colombo** through C072's `ScheduleLabels`, not in the handset's zone
 * (D-38): a driver naming yesterday's trip to support must name the same day support sees.
 */
@Composable
private fun tripLabel(trip: TripSummary): String {
    val locale = LocalConfiguration.current.locales[0]
    val route = listOfNotNull(trip.pickup?.address, trip.dropoff?.address)
        .joinToString(" ${ScheduleLabels.ROUTE_ARROW} ")
        .ifBlank { stringResource(R.string.support_trip_unnamed) }

    return "${ScheduleLabels.date(trip.startedAt, locale)} ${Symbols.DOT} $route"
}

private fun imageOnly() = PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly)

/** What the attachment is called in `docs.uploads`. Not user-facing. */
private const val SCREENSHOT_FILE_NAME = "support-screenshot.jpg"

/** The wireframe's `min-height:72px` description box, as lines. */
private const val MIN_DESCRIPTION_LINES = 3

File: apps/driver-ios/DriverApp/Support/SupportSheets.swift
import MageRideShared
import PhotosUI
import SwiftUI

/// SCR-DI-033's three overlays, hosted together because at most one is ever up.
///
/// The raise-ticket sheet is the wireframe's own **SCR-DI-033a**; the article and thread sheets are
/// US-16.1's *"+ detail"* and US-16.2's *"track ticket"*, which D2' names and the wireframe draws no
/// frame for. Both are `.sheet` rather than routes for that reason — adding two destinations the
/// team-approved baseline has no cell for would be a deviation, and a sheet over the list the driver
/// tapped is the least of it. All three take `.presentationDetents([.medium])`, which is D2'
/// §SCR-DI-033a's own SwiftUI clause.
@MainActor
struct SupportSheets: View {

    let sheet: SupportSheet
    @ObservedObject var model: SupportModel

    var body: some View {
        switch sheet {
        case .raiseTicket:
            RaiseTicketSheet(model: model)
        case .ticketThread:
            TicketThreadSheet(ticket: model.state.ticket)
        case .faqArticle:
            ArticleSheet(article: model.state.article)
        }
    }
}

/// **SCR-DI-033a · raise a ticket** — a `.sheet` at detent `.medium` (US-16.2).
///
/// The wireframe's four controls, in its order: **Issue description**, the **Related trip** picker,
/// *"📎 Attach a screenshot"* and **Submit ticket**. The title changes with the category — the same
/// sheet is the daily-fee refund request (US-9.23) — because both post the same operation and only
/// `category` differs.
@MainActor
private struct RaiseTicketSheet: View {

    @ObservedObject var model: SupportModel

    /// The system photo picker: **no `NSPhotoLibraryUsageDescription`**, because `PhotosPicker` is
    /// PHPicker and runs out of process, granting access to the one image the driver chose. The same
    /// contract SCR-DI-005's gallery fallback uses, and the same reason `Info.plist` deliberately
    /// carries no photo-library purpose string.
    @State private var picked: PhotosPickerItem?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                Text(key: model.state.isRefundRequest ? "support_refund_title" : "support_raise_ticket")
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)

                MultilineTextField(
                    labelKey: "support_issue_label",
                    placeholderKey: "support_issue_hint",
                    value: Binding(get: { model.state.description }, set: model.onDescriptionChange)
                )

                SectionLabel(key: "support_related_trip")
                tripPicker

                attachButton

                Button { Task { await model.submit() } } label: {
                    Text(key: "support_submit_ticket")
                }
                .buttonStyle(.mageCta(loading: model.state.isSubmitting))
                .disabled(!model.state.canSubmit)

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.background)
        .presentationDetents([.medium, .large])
        .onChange(of: picked) { item in
            Task { await load(item) }
        }
    }

    /// The wireframe's *"DRV-22011-0617 · Galle Face → Nugegoda ▾"*.
    ///
    /// **Δ iOS — a `Picker`, which is what D2' §SCR-DI-033a spells out** (*"past Trip ID `Picker`"*)
    /// where the Android twin builds an `ExposedDropdownMenuBox`. `.menu` style, so it reads as the
    /// wireframe's `▾` field rather than as a wheel; the id it shows is the route and the day, for
    /// ``SupportLabels/trip(_:)``'s reason.
    ///
    /// Optional, and *"No trip"* is a real option rather than a placeholder: a fee charged when the
    /// app crashed on Go Online is not about a trip at all, which is the commonest ticket this sheet
    /// opens.
    private var tripPicker: some View {
        Picker(
            selection: Binding(get: { model.state.tripId }, set: model.onTripSelected),
            label: Text(key: "support_related_trip")
        ) {
            Text(key: "support_trip_none").tag(String?.none)
            ForEach(model.state.trips, id: \.tripId) { trip in
                Text(SupportLabels.trip(trip)).tag(String?.some(trip.tripId))
            }
        }
        .pickerStyle(.menu)
        .labelsHidden()
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, MageRideSpacing.sm)
        .frame(minHeight: MageRideControl.minimumTapTarget)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }

    /// *"📎 Attach a screenshot"*, and what it says once one is attached.
    private var attachButton: some View {
        PhotosPicker(selection: $picked, matching: .images, photoLibrary: .shared()) {
            HStack(spacing: MageRideSpacing.xxs) {
                Image(systemName: model.state.screenshot == nil ? "paperclip" : "checkmark.circle.fill")
                    .font(.footnote)
                Text(key: model.state.screenshot == nil ? "support_attach" : "support_attached")
                    .mageFont(.subtitle)
            }
            .foregroundStyle(model.state.screenshot == nil ? MageRideColor.primary : MageRideColor.success)
            .frame(maxWidth: .infinity, minHeight: MageRideControl.ctaHeight)
            .overlay {
                RoundedRectangle(cornerRadius: MageRideControl.ctaRadius, style: .continuous)
                    .strokeBorder(MageRideColor.outline, lineWidth: 1)
            }
            .contentShape(Rectangle())
        }
    }

    /// Reads the picked image into memory.
    ///
    /// Bytes rather than the picker's item, for ``CapturedImage``'s own reason: a `PhotosPickerItem`
    /// is valid for the session that produced it, and this one has to survive until Submit.
    ///
    /// **`CaptureSource.gallery`**, which is honest and is also the only value that could be right: a
    /// screenshot is picked, never scanned. Nothing downstream reads it — `POST
    /// /v1/support/screenshots` declares no provenance part — and it is stamped anyway because
    /// ``CapturedImage`` carries provenance *with* the bytes by construction (AL-43).
    private func load(_ item: PhotosPickerItem?) async {
        guard let item, let data = try? await item.loadTransferable(type: Data.self) else { return }
        model.onScreenshotPicked(
            CapturedImage(
                fileName: Self.screenshotFileName,
                data: data,
                capturedVia: CaptureSource.gallery
            )
        )
    }

    /// What the attachment is called in `docs.uploads`. Not user-facing.
    private static let screenshotFileName = "support-screenshot.jpg"
}

/// One ticket and its whole conversation (US-16.2).
@MainActor
private struct TicketThreadSheet: View {

    let ticket: TicketDetail?

    var body: some View {
        ScrollView {
            Group {
                if let ticket {
                    thread(of: ticket)
                } else {
                    Text(key: "support_loading")
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.background)
        .presentationDetents([.medium, .large])
    }

    private func thread(of ticket: TicketDetail) -> some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            HStack(spacing: MageRideSpacing.xs) {
                Text(SupportLabels.category(ticket.category))
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)
                Spacer(minLength: MageRideSpacing.xs)
                StatusPill(
                    label: SupportLabels.statusKey(ticket.status).localised,
                    tone: SupportLabels.tone(ticket.status)
                )
            }

            // `IosTicketKt.ticketDescription`, not `ticket.description`: that name is `NSObject`'s on
            // this bridge and the Kotlin one is mangled around it — see the helper's own note.
            Text(IosTicketKt.ticketDescription(ticket: ticket))
                .mageFont(.body)
                .foregroundStyle(MageRideColor.onSurface)

            // Oldest first, as the contract sends it: a thread read bottom-up is a thread nobody
            // reads. `assigned` entries are skipped — who is handling a complaint is not the
            // driver's to see, which is the contract's own rule rather than this screen's.
            //
            // Keyed on the position, because a `TicketEvent` carries no id and two `responded`
            // entries a minute apart are genuinely two rows. Indices rather than
            // `Array(_:enumerated())`: **Swift has no key paths into tuples** (the C087 finding).
            ForEach(ticket.thread.indices, id: \.self) { index in
                threadEntry(ticket.thread[index])
            }
        }
    }

    /// One `TicketEvent` — what happened, and the agent's words when there are any.
    @ViewBuilder
    private func threadEntry(_ event: TicketEvent) -> some View {
        if let labelKey = SupportLabels.eventKey(event.kind) {
            VStack(alignment: .leading, spacing: 1) {
                Text(key: labelKey)
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)
                if let body = event.body, !body.isEmpty {
                    Text(body)
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurface)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
}

/// One FAQ article's body (US-16.1). Server-rendered in the driver's own language (D-26).
@MainActor
private struct ArticleSheet: View {

    let article: FaqArticle?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
                Text(article?.title ?? "support_loading".localised)
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)

                if let body = article?.body {
                    // Markdown as written. The app ships no renderer and support-svc's articles are
                    // short prose; a half-implemented one that swallowed `#` would be worse than
                    // none. `Text(_:)`'s own AttributedString markdown handles neither headings nor
                    // lists, so it is deliberately not reached for either.
                    Text(body)
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.background)
        .presentationDetents([.medium, .large])
    }
}
